lib/pipeline.js

```diff
@@ -1658,6 +1658,75 @@ class PipelineModel extends BaseModel {
         return factory.list(listConfig);
     }
 
+    /**
+     * Fetch builds belonging to a pipeline or events
+     * @param  {Object}   [config]
+     * @param  {Number}   [config.sort]                         Sort rangekey by ascending or descending
+     * @param  {Object}   [config.paginate]                     Pagination parameters
+     * @param  {Number}   [config.paginate.count]               Number of items per page
+     * @param  {Number}   [config.paginate.page]                Specific page of the set to return
+     * @param  {Number}   [config.groupEventId]                 Group event ID
+     * @param  {Boolean}  [config.latest]                       If we want to return latest in groupEventId, default false
+     * @return {Promise}  Resolves to an array of builds
+     */
+    async getBuilds(config) {
+        const latest = hoek.reach(config, 'params.latest');
+        const groupEventId = hoek.reach(config, 'params.groupEventId');
+
+        // Fetch all builds for each event with same groupEventId
+        if (groupEventId && !latest) {
+            // Lazy load factory dependency to prevent circular dependency issues
+            // https://nodejs.org/api/modules.html#modules_cycles
+            /* eslint-disable global-require */
+            const EventFactory = require('./eventFactory');
+            /* eslint-enable global-require */
+            const eventFactory = EventFactory.getInstance();
+
+            const events = await eventFactory.list({
+                params: { groupEventId }
+            });
+
+            if (Array.isArray(events) && events.length > 0) {
+                const processed = [];
+
+                events.forEach(e => processed.push(e.getBuilds()));
+
+                return Promise.all(processed).then(builds => {
+                    // flatten array of arrays
+                    return builds.flat(1);
+                });
+            }
+
+            return events;
+        }
+        // Lazy load factory dependency to prevent circular dependency issues
+        // https://nodejs.org/api/modules.html#modules_cycles
+        /* eslint-disable global-require */
+        const BuildFactory = require('./buildFactory');
+        /* eslint-enable global-require */
+        const buildFactory = BuildFactory.getInstance();
+
+        // Fetch only latest builds with same groupEventId
+        if (latest && groupEventId) {
+            return buildFactory.getLatestBuilds({
+                groupEventId: config.params.groupEventId
+            });
+        }
+
+        // Latest should not be passed to list config
+        if (latest !== undefined) {
+            delete config.params.latest;
+        }
+
+        // Fetch builds for this pipeline
+        const defaultConfig = {
+            params: { pipelineId: this.id }
+        };
+        const listConfig = config ? hoek.applyToDefaults(defaultConfig, config) : defaultConfig;
+
+        return buildFactory.list(listConfig);
```

Comment:
Do we keep a retention period for historical builds? I am not sure how much data could be returned as a response for builds of a given pipeline, but I am leaning towards not implementing this relaxation for now. Meaning without the supported params (in this case, it is only `groupEventId` for now), the client will be returned `400 bad request`.

Reply:
Added default pagination